Maxxor.PCL/ValueObject/MxSouthAfricanIdNumber/MxSouthAfricanIdNumber.cs

```diff
@@ -0,0 +1,178 @@
+                var month = int.Parse(monthSubString);
+                var year = int.Parse(yearSubString);
+
+                if (year > 0 && year < 11)
+                {
+                    year += 2000;
+                }
+                else
+                {
+                    year += 1900;
+                }
+
+                DateOfBirth = new MxDate(day, month, year);
+            }
+            catch (Exception e)
+            {
+                var fail = MxResult.Fail(this, MxSouthAfricanIdNumberError.InvalidDateOfBirth, e);
+                fail.Error.AddData("idNumber", idNumberString);
+                return fail;
+            }
+
+            return MxResult.Ok();
+        }
+
+        #endregion
+
+        #region Gender
+
+        public MxGender Gender { get; private set; }
+
+        private MxResult PopulateGender(string idNumberString)
+        {
+            var genderString = idNumberString.Substring(6, 4);
+            var genderNumber = int.Parse(genderString);
+
+            if (genderNumber >= 0 && genderNumber < 5000)
+            {
+                Gender = MxGender.Female;
+                return MxResult.Ok();
+            }
+            if (genderNumber >= 5000 && genderNumber <= 9999)
+            {
+                Gender = MxGender.Male;
+                return MxResult.Ok();
+            }
+
+            return MxResult.Fail(this, MxSouthAfricanIdNumberError.InvalidGender)
+                .AddData("idNumber", idNumberString); ;
+
+        }
+
+        #endregion
+
+        #region Citizenship
+
+        public bool IsCitizen { get; private set; }
+        public bool IsPermanentResident { get; private set; }
+
+        private MxResult PopulateCitizenship(string idNumberString)
+        {
+            var citizenshipSubstring = idNumberString.Substring(10, 1);
+
+            if (citizenshipSubstring == "0")
+            {
+                IsCitizen = true;
+                IsPermanentResident = false;
+            }
+            else
+            {
+                IsCitizen = false;
+                IsPermanentResident = true;
+            }
+
+            return MxResult.Ok();
+        }
+
+        #endregion
+
+        #region Checksum
+
+        private MxResult ValidateChecksum(string idNumberString)
+        {
+            var checksumString = idNumberString.Substring(12, 1);
+            var actualChecksum = int.Parse(checksumString);
+            var expectedChecksum = GetChecksumDigit(idNumberString);
+
+            if (actualChecksum == expectedChecksum)
+            {
+                IdNumberString = idNumberString;
+                return MxResult.Ok();
+            }
+
+            return MxResult.Fail(this, MxSouthAfricanIdNumberError.InvalidChecksum)
+                .AddData("idNumber", idNumberString)
+                .AddData("expectedChecksum", expectedChecksum.ToString());
+        }
+
+        public int GetChecksumDigit(string idNumberString)
```

Comment:
I believe this is the [Luhn algoritm](https://en.wikipedia.org/wiki/Luhn_algorithm)
In that case it works but it could be simpler, perhaps
```
var idCopy = string.Copy(idNumberString.Substring(0, 12)).Select(x => int.Parse(x.ToString())).ToArray();
for (var i = 0; i < 6; i++) 
{
    var num = idCopy[2 * i + 1] * 2;
    num = num > 9 ? num - 9 : num;
    idCopy[2 * i + 1] = num;
}
var sum = idCopy.Sum();
return 10 - (sum  % 10);
```

Reply:
The algorithm is elegant but string.Copy does not appear to work in a PCL. Also I made mine intentionally ugly so that some of the logic of the algorithm would be described in variable names.